File: src/services/inbox.ts
import type { Env, Viewer } from "../types";
import { json, readJson, unixNow } from "../utils";

interface InboxMessage {
  type: string;
  memoUid?: string;
  commentUid?: string;
  [key: string]: unknown;
}

interface DbInboxRow {
  id: number;
  created_ts: number;
  sender_id: number | null;
  receiver_id: number;
  status: "UNREAD" | "READ";
  message: string;
  sender_username?: string;
  sender_nickname?: string;
}

export async function listInbox(env: Env, viewer: Viewer): Promise<Response> {
  const rows = await env.DB.prepare(`
    SELECT inbox.*, "user".username AS sender_username, "user".nickname AS sender_nickname
    FROM inbox
    LEFT JOIN "user" ON "user".id = inbox.sender_id
    WHERE inbox.receiver_id = ?
    ORDER BY inbox.created_ts DESC
    LIMIT 100
  `).bind(viewer.id).all<DbInboxRow>();

  const unreadCount = await env.DB.prepare(`
    SELECT COUNT(*) AS count FROM inbox WHERE receiver_id = ? AND status = 'UNREAD'
  `).bind(viewer.id).first<{ count: number }>();

  return json({
    inbox: rows.results.map(row => ({
      id: row.id,
      createdTs: row.created_ts,
      sender: row.sender_id ? { id: row.sender_id, username: row.sender_username, nickname: row.sender_nickname } : null,
      status: row.status,
      message: safeParseMessage(row.message)
    })),
    unreadCount: unreadCount?.count ?? 0
  });
}

export async function updateInboxStatus(request: Request, env: Env, viewer: Viewer): Promise<Response> {
  const body = await readJson<{ ids?: number[]; status?: string }>(request);
  const status = body.status === "READ" ? "READ" : "UNREAD";

  if (body.ids?.length) {
    const placeholders = body.ids.map(() => "?").join(",");
    await env.DB.prepare(`
      UPDATE inbox SET status = ? WHERE id IN (${placeholders}) AND receiver_id = ?
    `).bind(status, ...body.ids, viewer.id).run();
  } else {
    await env.DB.prepare(`
      UPDATE inbox SET status = ? WHERE receiver_id = ?
    `).bind(status, viewer.id).run();
  }

  return json({ ok: true });
}

export async function deleteInboxItem(env: Env, viewer: Viewer, itemId: string): Promise<Response> {
  const id = Number(itemId);
  if (!Number.isFinite(id)) return json({ error: "Invalid inbox ID" }, 400);

  await env.DB.prepare("DELETE FROM inbox WHERE id = ? AND receiver_id = ?").bind(id, viewer.id).run();
  return json({ ok: true });
}

function safeParseMessage(value: string): InboxMessage {
  try {
    return JSON.parse(value) as InboxMessage;
  } catch {
    return { type: "unknown" };
  }
}

File: src/inbox.rs
use super::*;

pub(crate) async fn list_inbox(
    env: &Env,
    viewer: &Viewer,
) -> std::result::Result<Response, AppError> {
    let rows = db(env)?.prepare("SELECT inbox.*, \"user\".username AS sender_username, \"user\".nickname AS sender_nickname FROM inbox LEFT JOIN \"user\" ON \"user\".id = inbox.sender_id WHERE inbox.receiver_id = ? ORDER BY inbox.created_ts DESC LIMIT 100")
        .bind(&[js_num(viewer.id)])?
        .all()
        .await?;
    let inbox: Vec<DbInboxRow> = rows.results()?;
    let unread_count: Option<i64> = db(env)?
        .prepare("SELECT COUNT(*) AS count FROM inbox WHERE receiver_id = ? AND status = 'UNREAD'")
        .bind(&[js_num(viewer.id)])?
        .first(Some("count"))
        .await?;
    let payload: Vec<Value> = inbox.into_iter().map(public_inbox_item).collect();
    json_response(
        json!({ "inbox": payload, "unreadCount": unread_count.unwrap_or(0) }),
        200,
    )
    .map_err(AppError::from)
}

pub(crate) async fn update_inbox_status(
    req: &mut Request,
    env: &Env,
    viewer: &Viewer,
) -> std::result::Result<Response, AppError> {
    let body: Value = req.json().await.unwrap_or_else(|_| json!({}));
    let status = if body.get("status").and_then(Value::as_str) == Some("READ") {
        "READ"
    } else {
        "UNREAD"
    };
    let ids: Vec<i64> = body
        .get("ids")
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(Value::as_i64)
                .filter(|id| *id > 0)
                .collect()
        })
        .unwrap_or_default();

    if ids.is_empty() {
        db(env)?
            .prepare("UPDATE inbox SET status = ? WHERE receiver_id = ?")
            .bind(&[status.into(), js_num(viewer.id)])?
            .run()
            .await?;
    } else {
        let mut values: Vec<JsValue> = vec![status.into()];
        values.extend(ids.iter().map(|id| js_num(*id)));
        values.push(js_num(viewer.id));
        db(env)?
            .prepare(format!(
                "UPDATE inbox SET status = ? WHERE id IN ({}) AND receiver_id = ?",
                placeholders(ids.len())
            ))
            .bind(&values)?
            .run()
            .await?;
    }

    json_response(json!({ "ok": true }), 200).map_err(AppError::from)
}

pub(crate) async fn delete_inbox_item(
    env: &Env,
    viewer: &Viewer,
    item_id: &str,
) -> std::result::Result<Response, AppError> {
    let id = item_id
        .parse::<i64>()
        .map_err(|_| AppError::new(400, "Invalid inbox ID"))?;
    db(env)?
        .prepare("DELETE FROM inbox WHERE id = ? AND receiver_id = ?")
        .bind(&[js_num(id), js_num(viewer.id)])?
        .run()
        .await?;
    json_response(json!({ "ok": true }), 200).map_err(AppError::from)
}

pub(crate) fn public_inbox_item(row: DbInboxRow) -> Value {
    let sender = row.sender_id
        .map(|id| json!({ "id": id, "username": row.sender_username, "nickname": row.sender_nickname }))
        .unwrap_or(Value::Null);
    json!({
        "id": row.id,
        "createdTs": row.created_ts,
        "sender": sender,
        "status": row.status,
        "message": safe_inbox_message(&row.message)
    })
}

pub(crate) async fn record_comment_inbox(
    env: &Env,
    sender_id: i64,
    receiver_id: i64,
    parent_uid: &str,
    comment_uid: &str,
) -> std::result::Result<(), AppError> {
    db(env)?.prepare("INSERT INTO inbox (created_ts, sender_id, receiver_id, status, message) VALUES (?, ?, ?, 'UNREAD', ?)")
        .bind(&[
            js_num(unix_now()),
            js_num(sender_id),
            js_num(receiver_id),
            comment_inbox_message(parent_uid, comment_uid).to_string().into(),
        ])?
        .run()
        .await?;
    Ok(())
}

pub(crate) fn comment_inbox_message(parent_uid: &str, comment_uid: &str) -> Value {
    json!({
        "type": "memo.comment.created",
        "memoUid": parent_uid,
        "commentUid": comment_uid
    })
}

pub(crate) fn safe_inbox_message(value: &str) -> Value {
    serde_json::from_str::<Value>(value).unwrap_or_else(|_| json!({ "type": "unknown" }))
}
